text/3579-pki.md

```diff
@@ -0,0 +1,309 @@
+- Feature Name: Rust Public Key Infrastructure (PKI)
+- Start Date: 2024-02-27
+- RFC PR: [rust-lang/rfcs#3579](https://github.com/rust-lang/rfcs/pull/3579)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a plan and deployment for generating an initial set of Rust Project root and delegated cryptographic public/private key pairs, with associated X.509 certificate chains, to provide the proper hierarchy for deploying various signature validation techniques across the project.
+
+Having an internal Certificate Authority and delegate keys for the Rust Project allows us to have the infrastructure in place for solving multiple problems currently in the ecosystem such as: signing mirrors, crate package signing and validation, signing binaries for OS security checks, infrastructure security, and future chains of trust for authors and organizations.
+
+We propose the initial creation of a Rust Project Public Key hierarchy, with a root certificate authority and an initial set of single-purpose delegated certificates. For safety and oversight, this infrastructure mandates a quorum scheme (5-of-9) for root key operations, and mandates the use of Certificate Transparency for all subkey issuance *and* individual signature operations.
+
+Finally, we propose an initial, basic architecture for using this public key infrastructure to create delegate keys, as well as signing additional keys for usage in other signing schemas and implementations. Note, these additional implementations and further work is out of scope of this RFC. This RFC strictly covers creating a singular, root Certificate Authority and delegate keys for the Rust Project.
+
+## Out of Scope
+
+This RFC does not attempt to solve:
+- Identities or signing of Rust Project members
+- Identities or signing of crate authors
+- Answering open questions around the benefits or capabilities of any next-generation repository signing solutions (e.g. sigstore or the update framework (TUF))
+- Specific design or implementation strategies around binary signing (e.g. for Windows or macOS binaries)
+
+# Motivation
+[motivation]: #motivation
+
+The Rust Project currently lacks any official Public Key Infrastructure (PKI) required to perform proper package, mirror, code, and release signing across the project. There is high demand for capabilities that require such PKI infrastructure, such as mirroring and binary signing. This RFC provides the necessary cryptographic infrastructure to unblock such work, and an initial usage for cryptographic verification of Rust releases and `crates.io` downloads.
```

Comment:
I think this is not quite true. We do sign git tags and Rust releases with a keypair today; that isn't used for validation and the keypair isn't particularly secure (it has been on multiple machines and laptops and servers), but it is *something*.

Reply:
> We do sign git tags and Rust releases with a keypair today; that isn't used for validation and the keypair isn't particularly secure (it has been on multiple machines and laptops and servers), but it is _something_.

hi @Mark-Simulacrum , what do you mean by `that isn't used for validation`? From my understanding, the `*.asc` file in `rustup` is for users to valite the integrity of Rust release packages

Reply:
Rustup doesn't validate the pgp signature anymore: https://github.com/rust-lang/rustup/pull/3277 It has never been a hard error anyway, the validation code had been incorrectly rejecting older rustc versions (https://github.com/rust-lang/rustup/issues/3250) and it is expected to be replaced by a new system which this RFC lies the foundations for.